internal/runners/activate/activate.go

```diff
@@ -1,116 +1,136 @@
 package activate
 
 import (
-	"fmt"
 	"path/filepath"
 
+	"github.com/spf13/viper"
+
 	"github.com/ActiveState/cli/internal/analytics"
-	"github.com/ActiveState/cli/internal/constants"
+	"github.com/ActiveState/cli/internal/errs"
 	"github.com/ActiveState/cli/internal/failures"
+	"github.com/ActiveState/cli/internal/globaldefault"
 	"github.com/ActiveState/cli/internal/locale"
 	"github.com/ActiveState/cli/internal/logging"
 	"github.com/ActiveState/cli/internal/output"
 	"github.com/ActiveState/cli/internal/primer"
+	"github.com/ActiveState/cli/internal/runbits"
 	"github.com/ActiveState/cli/internal/subshell"
-	"github.com/ActiveState/cli/internal/virtualenvironment"
+	"github.com/ActiveState/cli/internal/updater"
+	"github.com/ActiveState/cli/pkg/cmdlets/git"
+	"github.com/ActiveState/cli/pkg/platform/model"
+	"github.com/ActiveState/cli/pkg/platform/runtime"
 	"github.com/ActiveState/cli/pkg/project"
 	"github.com/ActiveState/cli/pkg/projectfile"
 )
 
 type Activate struct {
-	namespaceSelect  namespaceSelectAble
-	activateCheckout CheckoutAble
+	namespaceSelect  *NamespaceSelect
+	activateCheckout *Checkout
 	out              output.Outputer
+	config           configAble
 	subshell         subshell.SubShell
 }
 
 type ActivateParams struct {
 	Namespace     *project.Namespaced
 	PreferredPath string
 	Command       string
+	Default       bool
 }
 
 type primeable interface {
 	primer.Outputer
 	primer.Subsheller
+	primer.Prompter
 }
 
-func NewActivate(prime primeable, namespaceSelect namespaceSelectAble, activateCheckout CheckoutAble) *Activate {
+func NewActivate(prime primeable) *Activate {
 	return &Activate{
-		namespaceSelect,
-		activateCheckout,
+		NewNamespaceSelect(viper.GetViper(), prime),
+		NewCheckout(git.NewRepo(), prime),
 		prime.Output(),
+		viper.GetViper(),
 		prime.Subshell(),
 	}
 }
 
 func (r *Activate) Run(params *ActivateParams) error {
-	return r.run(params, activationLoop)
+	return r.run(params)
 }
 
-func (r *Activate) run(params *ActivateParams, activatorLoop activationLoopFunc) error {
+func (r *Activate) run(params *ActivateParams) error {
 	logging.Debug("Activate %v, %v", params.Namespace, params.PreferredPath)
 
+	// Detect target path
 	pathToUse, err := r.pathToUse(params.Namespace.String(), params.PreferredPath)
 	if err != nil {
 		return locale.WrapError(err, "err_activate_pathtouse", "Could not figure out what path to use.")
 	}
 
-	projectToUse, err := r.projectToUse(pathToUse)
+	// Detect target project
+	proj, err := r.projectToUse(pathToUse)
 	if err != nil {
 		return locale.WrapError(err, "err_activate_projecttouse", "Could not figure out what project to use.")
 	}
 
 	// Run checkout if no project was given
-	if projectToUse == nil {
-		if !params.Namespace.IsValid() {
-			return locale.WrapError(err, "err_namespace_invalid", "Invalid namespace: {{.V0}}.", params.Namespace.String())
+	if proj == nil {
+		if params.Namespace == nil {
```

Comment:
This condition actually can never be true. It is always initialized: https://github.com/ActiveState/cli/blob/2b34f6b82e208668ea493c9a1fb432d7ffc0483a/cmd/state/internal/cmdtree/activate.go#L15
The `!params.Namespace.IsValid()` was correct.